File: 백준/Silver/2468. 안전 영역/안전 영역.cc
#include <iostream>
#include <queue>
using namespace std;

int arr[101][101];
int visited[101][101];

int maxi=0;
int n;
queue<pair<int,int>>q;
int dx[4]={0,0,1,-1};
int dy[4]={1,-1,0,0};


void bfs(int height, int tmp[101][101]){
  int cnt=0;
  for(int i=0;i<n;i++){
    for(int j=0;j<n;j++){
      visited[i][j]=0;
    }
  }
  for(int i=0;i<n;i++){
    for(int j=0;j<n;j++){
      if(arr[i][j]<=height){
        tmp[i][j]=0;
      }
      else tmp[i][j]=1;
    }
  }

  
  for(int i=0;i<n;i++){
    for(int j=0;j<n;j++){
      if(tmp[i][j]==1 && visited[i][j]==0){
        cnt++;
        q.push({i,j});
        visited[i][j]=1;
        while(!q.empty()){
          int front=q.front().first;
          int back=q.front().second;
          q.pop();
          for(int k=0;k<4;k++){
            int nx=front+dx[k];
            int ny=back+dy[k];

            if(nx>=0 && ny>=0 && nx<n && ny<n && tmp[nx][ny]==1 && visited[nx][ny]==0){
              q.push({nx,ny});
              visited[nx][ny]=1;
            }
          }
        }
      }
    }
  }
  maxi=max(maxi,cnt);
  return;
}
int main(){
  cin>>n;

  for(int i=0;i<n;i++){
    for(int j=0;j<n;j++){
      cin>>arr[i][j];
    }
  }

  for(int i=0;i<=100;i++){
    int tmp[101][101];
    bfs(i, tmp);
  }
  cout<<maxi;
}

File: 백준/Silver/2468. 안전 영역/안전 영역.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;

public class Main {
	public static int[] dx = {1,-1,0,0};
	public static int[] dy = {0,0,1,-1};
	public static int maxiCnt = 0;
	public static void main(String[] args) throws Exception{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int n = Integer.parseInt(br.readLine());
		int maxi = 0;

		int[][] arr = new int[n][n];
		int[][] tempArr = new int[n][n];
		for(int j=0;j<n;j++) {
			String[] tmp = br.readLine().split(" ");
			for(int k=0;k<tmp.length;k++) {
				arr[j][k] = Integer.parseInt(tmp[k]);
//				System.out.println(tmp[k]);
				maxi = Math.max(maxi, arr[j][k]);
			}
		}
		for(int i=0;i<=maxi;i++) {
			Queue<Pair> q = new LinkedList<Pair>();
			boolean[][] visited = new boolean[n][n];
			
			for(int j=0;j<n;j++) {
				for(int k=0;k<n;k++) {
					if(arr[j][k]<=i && !visited[j][k]) {
						tempArr[j][k] = 1;
						q.add(new Pair(j,k));
						visited[j][k] = true;
						while(!q.isEmpty()) {
							Pair p = q.poll();
							int x = p.x;
							int y = p.y;
							for(int l=0;l<4;l++) {
								int nx = x+dx[l];
								int ny = y+dy[l];
								if(nx>=0 && ny>=0 && ny<n && nx<n && !visited[nx][ny] && arr[nx][ny] <= i) {
									tempArr[nx][ny] = 1;
									visited[nx][ny] = true;
									q.add(new Pair(nx, ny));
								}
							}
						}
					}
				}
			}
//			for(int j=0;j<n;j++) {
//				for(int k=0;k<n;k++) {
//					System.out.print(tempArr[j][k]+" ");
//				}
//				System.out.println();
//			}
//			System.out.println();
			Queue<Pair> nq = new LinkedList<Pair>();

			int tmpCnt = 0;
			for(int j=0;j<n;j++) {
				for(int k=0;k<n;k++) {
					if(tempArr[j][k] == 0 && !visited[j][k]) {
						tmpCnt++;
						nq.add(new Pair(j,k));
						visited[j][k] = true;
						
						while(!nq.isEmpty()) {
							Pair p = nq.poll();
							int x = p.x;
							int y = p.y;
							for(int l=0;l<4;l++) {
								int nx = x+dx[l];
								int ny = y+dy[l];
								if(nx>=0 && ny>=0 && nx<n && ny<n && tempArr[nx][ny] == 0 && !visited[nx][ny]) {
									visited[nx][ny] = true;
									nq.add(new Pair(nx, ny));
								}
							}
						}
					}
				}
			}
			maxiCnt = Math.max(maxiCnt, tmpCnt);
		}
		System.out.println(maxiCnt);
	}
	
	public static class Pair{
		int x;
		int y;
		Pair(int x, int y){
			this.x =x;
			this.y =y;
		}
	}
}
